### DISCUSSION TITLE
How to extract data from Result type

### DISCUSSION THREAD

**Discussion by suraj-mandal** [gleam-lang/gleam#1985] [Questions & support] [answered]
How can I extract data from the **Result** datatype. I mean for example:

```elixir
import gleam/list.{last}
last_pascal_output = [[1], [1, 1]]
let last_row: Result(List(Int), Nil) = last(last_pascal_output)
```

This should return
```
Ok([1, 1])
```
Here how to extract `[1, 1]` from this?

**Comment by lpil** [comment, marked as answer]
Hello! There's a few different ways to do it.

The main way is to not extract the value, but instead to push the rest of the computation into the result. The `result` module is useful for this with it's functions `map`, `then`, etc.

```gleam
[[1], [1, 1]]
|> list.last
|> result.map(fn(x) {
  // do something with x here...
})
```
This can also be written with `use`

```gleam
use x <- result.map(list.last([[1], [1, 1]]))
// do something with x here...
```

Another way is to use `result.unwrap` or a case expression to extract the value out and use a default value if it is an `Error`.

Lastly you can use `assert Ok(x) = list.last(my_list)`. This is convenient but great care must be used as it will cause the program to crash if the pattern does not match.

**Reply by suraj-mandal** [reply]
Thanks a lot for your useful insights!
